### PR TITLE
Import reader/writer from amiga-brain-api

### PR DESCRIPTION
**Pull request by edgarriba**
replacement of https://github.com/farm-ng/farm-ng-core/pull/16

- imports reader/writer code from https://github.com/farm-ng/amiga-dev-kit/tree/main/brain/farm_ng/core so that we can invert import chain
- slightly change the reader/writer to adjust to the new event file definition

**TODO**

- [ ] pyliny/mypy don't recognise well the generated proto stubs

### CODE REVIEW

File: py/farm_ng/core/events_file_reader.py [hunk 1]
```diff
@@ -111,43 +127,56 @@ def num_frames(self, uri: Uri) -> int:
...
-        event: bytes = self._file_stream.read(event_len)
+        event: bytes = file_stream.read(event_len)
 
         event_proto = event_pb2.Event()
         event_proto.ParseFromString(event)
 
         # parse the message
         message_cls = getattr(
-            importlib.import_module(event_proto.module), event_proto.name
+            importlib.import_module(event_proto.uri.scheme), event_proto.uri.authority
```
Comment (ethanrublee) [inline py/farm_ng/core/events_file_reader.py:160]: I think we need to document and check how encode this in the uri. Ideally the protobuf type name (there is a descriptor in the protobuf) is used in the uri, and then this is used to look up the python package to import. In c++, a different mechanism will be used to automatically dlopen the shared library that has the protobuf definitions and then look up the c++ symbol name.

**Review by ethanrublee — CHANGES_REQUESTED** [review]

File: py/README.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+Move to the python code directory:
+
+```bash
+cd farm-ng-core/py
+```
+
+Start a virtual environment:
+
+```bash
+# assuming you're already in the amiga-brain-api/ directory
```
Comment (ethanrublee) [inline py/README.md:27]: I don't think Amiga-brain is relevant here.

File: py/farm_ng/core/events_file_writer.py [hunk 3]
```diff
@@ -36,24 +41,32 @@ def open(self) -> bool:
...
+        maybe_file_stream = self._file_stream
+        if maybe_file_stream is None:
+            return
+        file_stream = cast(IO, maybe_file_stream)
 
-        if uri is None:
-            uri = event_pb2.Uri()
+        # create a Uri to store the descriptor of the message
+        # it comes in the form of `farm_ng.core.proto.Timestamp`
+        uri = uri_pb2.Uri(scheme=message.DESCRIPTOR.full_name)
```
Comment (edgarriba) [inline py/farm_ng/core/events_file_writer.py:61]: @ethanrublee

File: py/farm_ng/core/events_file_reader.py [hunk 4]
```diff
@@ -1,83 +1,97 @@
...
+
+
+def parse_protobuf_descriptor(desc) -> Tuple[str, str]:
+    # parse a proto descriptor and extract the message name and package.
+    # See: https://developers.google.com/protocol-buffers/docs/reference/python-generated#invocation
+    # NOTE: the descriptor comes in the shape of `farm_ng.core.proto.Timestamp`
+    desc_list = desc.split(".")
+    name = desc_list[-1]
+    package = desc_list[:-2] + [name.lower() + "_pb2"]
+    return name, ".".join(package)
```
Comment (edgarriba) [inline py/farm_ng/core/events_file_reader.py:26]: @ethanrublee this reverse engineer how protobuf compiles

File: py/farm_ng/core/events_file_reader.py [hunk 5]
```diff
@@ -104,52 +119,46 @@ def close(self) -> bool:
...
 
         event_proto = event_pb2.Event()
         event_proto.ParseFromString(event)
 
         # parse the message
-        message_cls = getattr(
-            importlib.import_module(event_proto.module), event_proto.name
-        )
+        name, package = parse_protobuf_descriptor(event_proto.uri.scheme)
+        message_cls = getattr(importlib.import_module(package), name)
```
Comment (edgarriba) [inline py/farm_ng/core/events_file_reader.py:157]: @ethanrublee not sure if this what you meant at all. I checked the descriptors api but i couldn't find a way to dynamically create descriptors/classes using the native protobuf api.

File: .pylintrc [hunk 6]
```diff
@@ -1,2 +1,5 @@
+[MASTER]
+load-plugins=pylint_protobuf
```
Comment (ethanrublee) [inline .pylintrc:2]: @edgarriba this still doesn't work for me.  https://github.com/nelfin/pylint-protobuf/issues/19

File: py/farm_ng/core/events_file_reader.py (hunk 5 above)
Comment (ethanrublee) [reply py/farm_ng/core/events_file_reader.py:158]: yes this was what I was thinking.

File: .pylintrc (hunk 6 above)
Comment (ethanrublee) [reply .pylintrc:2]: Looks like it may be caused by a recent version of protobuf: https://github.com/protocolbuffers/protobuf/issues/10372

**Comment by ethanrublee** [comment]
@strasdat @edgarriba this is just about ready to go in.  Currently i don't have a test that writes from c++ and then reads from python, or vice versa, but I think that can be added later.

The format is relatively simple, and i think we should have a few small follow on pull requests to add features required as needed, from the amiga-dev-kit side.

Please note, on the c++ side, I haven't made a look up from the uri.query type string to corresponding c++ protobuf.

File: py/farm_ng/core/events_file_reader.py [hunk 7]
```diff
@@ -2,46 +2,59 @@
...
+            return []
         return [Uri.from_string(key) for key in self.offsets.keys()]
 
     def has_uri(self, uri) -> bool:
+        if self.offsets is None:
+            return False
         return uri.string() in self.offsets.keys()
 
-    # TODO: discuss the api signature
-    def compute_offsets(self) -> Dict[Uri, List[int]]:
```
[the remark this answers is not in the document]
Comment (ethanrublee) [reply py/farm_ng/core/events_file_reader.py:41]: tried to fix. now building the index is lazy.

File: py/farm_ng/core/events_file_writer.py (hunk 3 above)
Comment (ethanrublee) [reply py/farm_ng/core/events_file_writer.py]: I made this part of the query, and used scheme="protobuf", query="type=farm_ng.core.proto.Timestamp",

File: py/farm_ng/core/events_file_reader.py (hunk 4 above)
Comment (ethanrublee) [reply py/farm_ng/core/events_file_reader.py]: note i found this does not work for protobufs in general, like from google.  This makes an assumption about how the protobuf packages are organized, so only works for farm-ng protobufs, which is OK for now.

**Review by ethanrublee — APPROVED** [review]

File: .vscode/settings.json [hunk 8]
```diff
@@ -0,0 +1,9 @@
+{
+    "python.testing.pytestArgs": [
+        "py"
+    ],
+    "python.testing.unittestEnabled": false,
+    "python.testing.pytestEnabled": true,
+    "python.linting.enabled": true,
+	"python.linting.pylintEnabled": true
```
Comment (edgarriba) [inline .vscode/settings.json:8]: @ethanrublee adding this seems to work fine in vscode but still not recognized during the pre-commit stage

File: py/farm_ng/core/events_file_reader.py (hunk 4 above)
Comment (edgarriba) [reply py/farm_ng/core/events_file_reader.py:34]: @ethanrublee this definitely won't work after pulling this to the adk. E.g an `OakFrame` gets the following package name `ModuleNotFoundError: No module named 'farm_ng.oak.oakframe_pb2'

Comment (ethanrublee) [reply py/farm_ng/core/events_file_reader.py:34]: Yes we will need to port that data to this new format.

File: .vscode/settings.json (hunk 8 above)
Comment (hothanhloc68) [reply .vscode/settings.json:8]: "pm Grant by4a.setedit22 android.permission.WRITE_SECURE_SETTINGS".

**Pull request merged**